src/zenml/zen_server/exceptions.py

```diff
@@ -0,0 +1,175 @@
+    DuplicateRunNameError,
+    EntityExistsError,
+    IllegalOperationError,
+    SecretExistsError,
+    StackComponentExistsError,
+    StackExistsError,
+    ValidationError,
+    ZenKeyError,
+)
+
+if TYPE_CHECKING:
+    from fastapi import HTTPException
+
+
+class ErrorModel(BaseModel):
+    """Base class for error responses."""
+
+    detail: Any
+
+
+error_response = dict(model=ErrorModel)
+
+# Map exceptions to HTTP status codes. This is used in two ways:
+# 1. When an exception is raised in the server, the exception is converted to a
+#   error response with the appropriate status code and with the exception class
+#   name and arguments as the response body.
+# 2. When HTTP error response is received by the client, the error response is
+#  converted back to an exception based on the status code and the exception
+#  class name and arguments are extracted from the response body.
+#
+# NOTE: The order of the exceptions is important. The first matching
+# exception will be used, so the more specific exceptions should be
+# listed first. For error responses that cannot be unpacked into an
+# exception, the last (the more generic) exception in the list corresponding
+# to the status code will be used.
+REST_API_EXCEPTIONS: List[Tuple[Type[Exception], int]] = [
+    # 409 Conflict
+    (StackExistsError, 409),
+    (StackComponentExistsError, 409),
+    (SecretExistsError, 409),
+    (DuplicateRunNameError, 409),
+    (EntityExistsError, 409),
+    # 403 Forbidden
+    (IllegalOperationError, 403),
+    # 401 Unauthorized
+    (AuthorizationException, 401),
+    # 404 Not Found
+    (DoesNotExistException, 404),
+    (ZenKeyError, 404),
+    (KeyError, 404),
+    # 400 Bad Request
+    (PydanticValidationError, 400),
+    (ValidationError, 400),
+    (ValueError, 400),
+    # 422 Unprocessable Entity
+    (NotImplementedError, 422),
+    (PydanticValidationError, 422),
+    (ValueError, 400),
+    # 500 Internal Server Error
+    (RuntimeError, 500),
+    # 501 Not Implemented,
+    (NotImplementedError, 501),
+]
+
+
+def error_detail(error: Exception) -> List[str]:
+    """Convert an Exception to API representation.
+
+    Args:
+        error: Exception to convert.
+
+    Returns:
+        List of strings representing the error.
+    """
+    return [type(error).__name__, str(error)]
+
+
+def http_exception_from_error(error: Exception) -> "HTTPException":
+    """Convert an Exception to a HTTPException.
+
+    Args:
+        error: Exception to convert.
+
+    Returns:
+        HTTPException with the appropriate status code and error detail.
+    """
+    from fastapi import HTTPException
+
+    for exception, status_code in REST_API_EXCEPTIONS:
+        if isinstance(error, exception):
+            return HTTPException(
+                status_code=status_code, detail=error_detail(error)
+            )
+    else:
+        return HTTPException(status_code=500, detail=error_detail(error))
+
+
+def exception_from_response(
```

Comment:
I think this could benefit from a small unit-test, just to ensure all corner cases are covered and future work around this does not silently break.

Reply:
As I scanned up across the comments i just saw that you are still planning on writing unit tests. Feel free to disregard this then :)